### ISSUE TITLE
patch_get_utility problems
Search
￼

### ISSUE THREAD

**Issue by bmw** [certbot/certbot#4071]
```
In [2]: from certbot.tests.util import patch_get_utility

In [3]: with patch_get_utility():
   ...:     print zope.component.getUtility("foo")
   ...:     zope.component.getUtility.return_value = 42
   ...:     print zope.component.getUtility("foo")
   ...:     
<certbot.tests.util.FreezableMock object at 0x7f664a372450>
42
```
We shouldn't let people do that.

**Comment by bmw** [comment]
The purpose of `patch_get_utility()` is described in more detail in #4010. To give a short summary, we have a lot of places in the code where we do things like [this](https://github.com/certbot/certbot/blob/v0.14.1/certbot/main.py#L263). We use `zope.component.getUtility` to get a display object and then do some kind of user interaction.

The problem is that during our unit tests, these utilities and displays aren't set up so we mock the call. This is fine, however, there are a number of optional parameters to these display objects and some combinations aren't valid. By mocking the call, these assertions never run and we've had code that fails assertions land in `master` because the display object was never properly called because it was mocked out during tests.

`patch_get_utility` mocks out the `getUtility` call, but also runs the assertions on the parameters. The problem identified in the original post of this issue is there is a way `patch_get_utility` causes all of this special functionality to be ignored. This should be fixed.

**Comment by cj-dev** [comment]
I'm new here, so excuse me if I'm way off. As I understand the mock objects used for the display...

`tests.util.patch_get_utility` mocks the `getUtility` function by patching it with a call to `tests.util._create_get_utility_mock` which returns a `tests.util.FreezableMock`. The FreezableMock object itself doesn't have any assertions (its `_func` is None) but its attributes inherited from `interfaces.IDisplay` (other than 'notification') have their parameters checked by `display.util.assert_valid_call`, containing an assertion.

What you showed in this thread's original post, `zope.component.getUtility.return_value = 42`, is legitimate mock behavior, right? And from my understanding above, I'd not expect the assertion function to be called when assigning to `return_value`. Are you suggesting that the mocked-out `getUtility` shouldn't have a `return_value` attribute that can be set after-the-fact? Maybe I need to see examples of the other code you mention that fails assertions getting into `master`.

```
(Pdb) zope.component.getUtility('foo').yesno('yes or no', default='yes')
<MagicMock name='mock()' id='140186259122896'>
(Pdb) zope.component.getUtility('foo').yesno('yes or no')
*** AssertionError: Invalid IDisplay call for this prompt:
yes or no
(Pdb)
```

**Comment by bmw** [comment]
You're on the right track. Here's [the code](https://github.com/cowlicks/letsencrypt/blob/95c8ca98c0648a66f1e9c7984c68d7c8fdd31e58/certbot/tests/main_test.py#L1186) that prompted this issue. After switching to `tests.util.patch_get_utility`, that code still has to be modified. While using `return_value` in this way is totally valid, it inadvertently  prevents the assertions from running that prevent bugs like #3934 and #3996.

While this code is pretty ugly, it prevents regressions. I'm open to alternatives, but I think that erroring out with a nice error message for developers about how you shouldn't set `return_value` to another object here is probably the best approach to solving this issue.

**Comment by cj-dev** [comment]
Oh! That helps a ton. I'll get back to you soon.

**Comment by cj-dev** [comment]
Opened two PRs with candidate solutions (neither should be merged in their current states). I wanted to get a sense for how important preventing the `zope.component.getUtility.return_value = 42` case is. Because the patched callable is of type `mock.MagicMock`, I haven't found a way to prevent its `return_value` from being overwritten without using a subclass of MagicMock. Example for why that's not the easiest thing in the world is in #4963.

Wondering if either approach is worth following. The first approach is probably easier to implement, but doesn't really fix every instance of this issue; if the second, I'll take a while longer and I might have to concede I don't know the intersection of Python object internals and Mock well enough to be efficient on this issue. I will at least try to see either one through. I'm still actively researching other ways to address this.

**Comment by cj-dev** [comment]
Restored some confidence with the most recent commit to #4963 . Will definitely take another look with fresh eyes later today.

**Comment by bmw** [comment]
Fixed in #4963.
